design/one-pager-crd-scaling.md

```diff
@@ -0,0 +1,150 @@
+
+As the Crossplane community, what we would expect in these scenarios are:
+- No API service disruptions like the incidents described in [[4]] in cases where we have ~2000 CRDs installed.
+- It takes the discovery client no more than 10s to discover all available GVs. This would cover both the initial discovery needed by the controllers and periodic discoveries triggered by other clients such as `kubectl`. We propose a 10s goal because even if client-side throttling is no longer the bottleneck, it looks like that the REST client needs ~8s to fully consume responses (probably another investigation and improvement point).
+
+For the 1st scenario above, we are already in a good position [except](https://github.com/crossplane/crossplane/issues/2895) for GKE zonal and regional clusters as of Kubernetes `v1.23.1`, probably due to the initial control-plane sizing chosen in their control-planes. And regarding the client-side throttling issues, as discussed in the "Cient-side Throttling" section above, with Kubernetes `v1.24` the token bucket rate limiter used by the discovery client allows request bursts upto 300, which is well above a singe provider's (`provider-jet-aws@v0.4.0-preview`) GVs (~200 GVs). So we anticipate an improved UX with the upcoming Kubernetes `v1.24`.
+
+For the 2nd scenario above, we reach ~370 GVs (including some other "external" CRDs) in the cluster, which is above the allowed burstiness of even the `v1.24` discovery client. And again client-side throttling becomes a bottleneck if the discovery client needs to discover the whole API (empty cache, cache invalidation, etc.). Also even on an AKS `v1.22.4` cluster, we were not able to install `provider-jet-{aws,gcp,azure}` `preview` editions together.
+
+When the criterias discussed above are met for the Crossplane scenarios we currently anticipate, and as long as we can satisfy these criterias with future Crossplane or upstream changes, we should be in a good position.
+
+
+## Testing Against the Criteria Set
+
+One can use the simple Posix `time` utility to measure the time it takes for `kubectl` to just establish the discovery cache:
+```bash
+rm -fR  ~/.kube/cache/discovery/<host_port> && time (kubectl api-resources > /dev/null)
+```
+
+A more realistic example that also performs a final HTTP GET to fetch a `PodList` resource:
+```bash
+rm -fR  ~/.kube/cache/discovery/<host_port> && time (kubectl get pods > /dev/null)
+```
+
+## Action Items
+- We can consider cherry-picking [[5]] to all active release branches `v1.23`, `v1.22`, `v1.21`, as the anticipated release date for the `v1.24` release is in April, 2022.
+- Open issues regarding API service disruptions for managed control-planes (GKE regional, AKS, EKS, etc.), where we expect high-availability.
```

Comment:
Some folks on today's community meeting (namely @haarchri and Jillian Hill from Guidewire) mentioned that their AWS professional services teams have noticed increase resource usage even when EKS appears to be working well and reached out to investigate. It sounds like in some cases they're also seeing EKS clusters choke when installing providers with a lot of CRDs - one theory is that some regions have smaller EKS control planes.

Either way it sounds like we might be able to reach out either to our own contacts at AWS or through a company in the Crossplane community to get some insight into how EKS is handling this load.

Reply:
That would be great Nic. The support issue we opened for GKE regional clusters was not fruitful, the outcome was it works as expected. What we wanted to learn was basically the metrics GKE autoscalers are using when deciding to scale their managed control-planes up (CPU/memory usage/utilization of control-plane components, kube-apiserver or other component failures, # of CRDs, or some other SLIs, etc.?). One thing I still wonder is how [SLAs](https://cloud.google.com/kubernetes-engine/sla) come into play as Crossplane users install three big providers in their clusters. @AaronME, do you know if it's possible, in cloud consoles or by some other means (like a support ticket), we can inquiry about these SLAs?